Mage/src/main/java/mage/util/CardUtil.java

```diff
@@ -1148,6 +1155,116 @@ public static void makeCardPlayable(Game game, Ability source, Card card, Durati
+                spellAbilities.add(((ModalDoubleFacesCard) cardToCast).getRightHalfCard().getSpellAbility());
+            }
+        } else {
+            spellAbilities.add(cardToCast.getSpellAbility());
+        }
+        spellAbilities.removeIf(Objects::isNull);
+        return spellAbilities;
+    }
+
+    private static final FilterCard defaultFilter = new FilterCard("card to cast");
+
+    private static TargetCard getTarget(Set<UUID> cardIds, Game game) {
+        Zone zone = game.getState().getZone(cardIds.stream().findFirst().orElse(null));
+        switch (zone) {
+            case HAND:
+                return new TargetCardInHand(0, 1, defaultFilter);
+            case GRAVEYARD:
+                return new TargetCardInGraveyard(0, 1, defaultFilter);
+            case EXILED:
+                return new TargetCardInExile(0, 1, defaultFilter, null);
+            case LIBRARY:
+                return new TargetCardInLibrary(0, 1, defaultFilter);
+            default:
+                throw new UnsupportedOperationException("can't choose a card from zone: " + zone);
+        }
+    }
+
+    public static boolean castSpellWithAttributesForFree(Player player, Ability source, Game game, Cards cards, FilterCard filter) {
+        Map<UUID, List<SpellAbility>> cardMap = new HashMap<>();
+        for (Card card : cards.getCards(game)) {
+            List<SpellAbility> spellAbilities = getCastableWithAttributes(card, filter, source.getSourceId(), player.getId(), game);
+            if (!spellAbilities.isEmpty()) {
+                cardMap.put(card.getId(), spellAbilities);
+            }
+        }
+        Card cardToCast;
+        switch (cardMap.size()) {
+            case 0:
+                return false;
+            case 1:
+                cardToCast = cards.get(cardMap.keySet().stream().findFirst().orElse(null), game);
+                break;
+            default:
+                Cards castableCards = new CardsImpl(cardMap.keySet());
+                TargetCard target = getTarget(cardMap.keySet(), game);
+                player.choose(Outcome.PlayForFree, castableCards, target, game);
+                cardToCast = castableCards.get(target.getFirstTarget(), game);
+        }
+        if (cardToCast == null) {
+            return false;
+        }
+        List<SpellAbility> partsToCast = cardMap.get(cardToCast.getId());
+        String partsInfo = partsToCast.stream()
+                .map(SpellAbility::getName)
+                .collect(Collectors.joining(" or "));
+        if (cardToCast == null
+                || partsToCast.size() <= 0
+                || !player.chooseUse(
+                Outcome.PlayForFree, "Cast spell without paying its mana cost (" + partsInfo + ")?", source, game
+        )) {
+            return false;
+        }
+        game.getState().setValue("PlayFromNotOwnHandZone" + cardToCast.getId(), Boolean.TRUE);
```

Comment:
That's wrong. You must mark card parts as playable, not cards. See Cascade as example.